academic_observatory_workflows/workflows/orcid_telescope.py

```diff
@@ -0,0 +1,756 @@
+    gcs_list_blobs,
+    gcs_hmac_key,
+)
+from observatory.platform.observatory_config import CloudWorkspace
+from observatory.platform.workflows.workflow import (
+    Workflow,
+    ChangefileRelease,
+    cleanup,
+    set_task_state,
+)
+
+ORCID_AWS_SUMMARIES_BUCKET = "v2.0-summaries"
+ORCID_REGEX = r"\d{4}-\d{4}-\d{4}-\d{3}(\d|X)\b"
+ORCID_RECORD_REGEX = r"\d{4}-\d{4}-\d{4}-\d{3}(\d|X)\.xml$"
+MANIFEST_HEADER = ["bucket_name", "blob_name", "updated"]
+BATCH_REGEX = r"^\d{2}(\d|X)$"
+
+
+class OrcidBatch:
+    """Describes a single ORCID batch and its related files/folders"""
+
+    def __init__(self, download_dir: str, transform_dir: str, batch_str: str):
+        self.download_dir = download_dir
+        self.transform_dir = transform_dir
+        self.batch_str = batch_str
+        self.download_batch_dir = os.path.join(self.download_dir, batch_str)
+        self.download_log_file = os.path.join(self.download_dir, f"{self.batch_str}_log.txt")
+        self.download_error_file = os.path.join(self.download_dir, f"{self.batch_str}_error.txt")
+        self.manifest_file = os.path.join(self.download_dir, f"{self.batch_str}_manifest.csv")
+        self.transform_upsert_file = os.path.join(self.transform_dir, f"{self.batch_str}_upsert.jsonl.gz")
+        self.transform_delete_file = os.path.join(self.transform_dir, f"{self.batch_str}_delete.jsonl.gz")
+
+        if not os.path.exists(self.download_dir):
+            raise NotADirectoryError(f"Directory {self.download_dir} does not exist.")
+        if not os.path.exists(self.transform_dir):
+            raise NotADirectoryError(f"Directory {self.transform_dir} does not exist.")
+        if not re.match(BATCH_REGEX, self.batch_str):
+            raise ValueError(f"Batch string {self.batch_str} is not valid.")
+
+        os.makedirs(self.download_batch_dir, exist_ok=True)
+
+    @property
+    def expected_records(self) -> List[str]:
+        """List of expected ORCID records for this ORCID directory. Derived from the manifest file"""
+        with open(self.manifest_file, "r") as f:
+            reader = csv.DictReader(f)
+            return [os.path.basename(row["blob_name"]) for row in reader]
+
+    @property
+    def existing_records(self) -> List[str]:
+        """List of existing ORCID records on disk for this ORCID directory."""
+        return [os.path.basename(path) for path in list_files(self.download_batch_dir, ORCID_RECORD_REGEX)]
+
+    @property
+    def missing_records(self) -> List[str]:
+        """List of missing ORCID records on disk for this ORCID directory."""
+        return list(set(self.expected_records) - set(self.existing_records))
+
+    @property
+    def blob_uris(self) -> List[str]:
+        """List of blob URIs from the manifest this ORCID directory."""
+        with open(self.manifest_file, "r") as f:
+            reader = csv.DictReader(f)
+            return [gcs_blob_uri(bucket_name=row["bucket_name"], blob_name=row["blob_name"]) for row in reader]
```

Comment:
It may be beneficial to read the `manifest_file` in the constructor, so that it doesn't have to be read in each time a property is called.

Also, for a given `manifest_file`, do the results returned by each property ever change while that instance is alive? If not, you could use of `cached_property` from `from functools import cached_property` as this will cache the results if the property is called more than one time.

See: https://docs.python.org/3/library/functools.html#functools.cached_property

Reply:
I didn't know this cached_property existed, this is really good! 

The expected_records and blob_uris properties aren't expected to ever change for any given instance so they can be cached. However the other two should be calculated each time they're called. 

The manifest file doesn't necessarily exist on construction, so I'd rather not put it in the constructor as it'd introduce some ugly logic.

Reply:
Yeah, I haven't used it before either but should. OK fair enough.